tensorflow_addons/optimizers/novograd.py

```diff
@@ -0,0 +1,246 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Novograd for TensorFlow."""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+from tensorflow.python.training import training_ops
+
+
+@tf.keras.utils.register_keras_serializable(package='Addons')
+class Novograd(tf.keras.optimizers.Optimizer):
+    """The Novograd Optimizer was first proposed in [Stochastic Gradient
+    Methods with Layerwise Adaptvie Moments for training of Deep
+    Networks](https://arxiv.org/pdf/1905.11286.pdf)
+
+    NovoGrad is a first-order SGD-based algorithm, which computes second
+    moments per layer instead of per weight as in Adam. Compared to Adam,
+    NovoGrad takes less memory, and has been found to be more numerically
+    stable. More specifically we compute (for more information on the
+    computation please refer to this
+    [link](https://nvidia.github.io/OpenSeq2Seq/html/optimizers.html):
+
+    ```
+    Second order moment = exponential moving average of Layer-wise square
+    of grads:
+        v_t <-- beta_2 * v_{t-1} + (1-beta_2) * (g_t)^2
+    First order moment in one of four modes:
+        1. moment of grads normalized by v_t:
+            m_t <- beta_1 * m_{t-1} + [ g_t / (sqrt(v_t)+epsilon)]
+        2. moment similar to Adam: exponential moving average of grads
+        normalized by v_t (set grad_averaging = True to use this):
+            m_t <- beta_1 * m_{t-1} +
+                   [(1 - beta_1) * (g_t / (sqrt(v_t) + epsilon))]
+        3. weight decay adds a w_d term after grads are rescaled by
+        1/sqrt(v_t) (set weight_decay > 0 to use this0:
+            m_t <- beta_1 * m_{t-1} +
+                   [(g_t / (sqrt(v_t) + epsilon)) + (w_d * w_{t-1})]
+        4. weight decay + exponential moving average from Adam:
+            m_t <- beta_1 * m_{t-1} +
+                   [(1 - beta_1) * ((g_t / (sqrt(v_t + epsilon)) +
+                   (w_d * w_{t-1}))]
+    Weight update:
+        w_t <- w_{t-1} - lr_t * m_t
+    ```
+
+    Example of usage:
+    ```python
+    opt = tfa.optimizers.Novograd(
+        lr=1e-3,
+        beta_1=0.9,
+        beta_2=0.999,
+        weight_decay=0.001,
+        grad_averaging=False
+    )
+    ```
+    """
+
+    def __init__(self,
+                 learning_rate=0.001,
+                 beta_1=0.9,
+                 beta_2=0.999,
+                 epsilon=1e-8,
+                 weight_decay=0.0,
+                 grad_averaging=False,
```

Comment:
I find the default arguments are very different from the original implementation. Is it because we use adam instead of momentum optimizer (SGD with momentum)?

Reply:
The default arguments I've added in are more in line with the default arguments provided to Adam across different libraries. I didn't see any point in moving away from the convention of having `beta_1` and `beta_2` be 0.9 and 0.999 respectively. In my tests novoograd with 0.95 and 0.98 was not significantly better than novograd with 0.9 and 0.999 so I went with the convention.

Reply:
Got it! Thanks for the survey.

Reply:
@shreyashpatodia, one minor thing is that default `epsilon` for `tf.keras.optimizers.Adam` is `1e-7` instead of `1e-8`.

Reply:
Changed.